stats/cloud/logs.go

```diff
@@ -0,0 +1,176 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2020 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cloud
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/mailru/easyjson"
+	"github.com/sirupsen/logrus"
+)
+
+//easyjson:json
+type msg struct {
+	Streams        []msgStreams        `json:"streams"`
+	DroppedEntries []msgDroppedEntries `json:"dropped_entries"`
+}
+
+//easyjson:json
+type msgStreams struct {
+	Stream map[string]string `json:"stream"`
+	Values [][2]string       `json:"values"` // this can be optimized
+}
+
+//easyjson:json
+type msgDroppedEntries struct {
+	Labels    map[string]string `json:"labels"`
+	Timestamp string            `json:"timestamp"`
+}
+
+func (m *msg) Log(logger logrus.FieldLogger) {
+	var level string
+
+	for _, stream := range m.Streams {
```

Comment:
Hmm this could potentially cause messages to arrive out-of-order, right? I doubt it's a big issue, if each `msg` encapsulates only a second, since each stream would be a separate instance?

Reply:
I thought of this some minutes ago ... as `/query_range` apperantly doesn't return logs in any given order as well. I don't even know if tail does ... my testing - shows that - yes it does seem to come in order, but more investigation is required.

I fear I will need to get all the logs and then to order them and then print them :( But I propose that is a separate PR - this is unlikely to be a problem unless you have a lot of logs and arguably that probably is already pretty bad experience (also likely to lead to dropped logs due to k6 not being able to tail fast enough)

Reply:
:+1: fine for a separate PR, if needed...